src/server/ftpserver.rs

```diff
@@ -515,17 +561,35 @@ where
                     logger: self.logger.clone(),
                     options: (&self).into(),
                     shutdown_topic: shutdown_notifier.clone(),
+                    failed_logins: failed_logins.clone(),
                 }
                 .listen(),
             ) as Pin<Box<dyn Future<Output = std::result::Result<(), ServerError>> + Send>>,
         };
 
-        tokio::select! {
-            result = listen_future => result,
-            opts = self.shutdown => {
-                slog::debug!(logger, "Shutting down within {:?}", opts.grace_period);
-                shutdown_notifier.notify().await;
-                Self::shutdown_linger(logger, shutdown_notifier, opts.grace_period).await
+        // There must be a better way to do this?!
```

Comment:
See if this works:

```rust
        let sweeper = if let Some(ref failed_logins) = failed_logins {
            Box::pin(failed_logins.sweeper(self.logger.clone(), shutdown_notifier.clone())) as Pin<Box<dyn futures_util::Future<Output = ()>>>
        } else {
            Box::pin(futures_util::future::pending()) as Pin<Box<dyn futures_util::Future<Output = ()>>>
        };
        tokio::select! {
            result = listen_future => result,
            _ = sweeper => {
                Ok(())
            },
            opts = self.shutdown => {
                slog::debug!(logger, "Shutting down within {:?}", opts.grace_period);
                shutdown_notifier.notify().await;
                Self::shutdown_linger(logger, shutdown_notifier, opts.grace_period).await
            }
        }

```

Reply:
if there is no failed logins cache the select! statement's `sweeper` future will just simply never return.

Reply:
It works. Awesome! Was looking for this.

Reply:
Only doesn't work for the test which uses tokio::spawn and gives a trait bounds error 'is not Send'

Reply:
Try to cast to `as Pin<Box<dyn futures_util::Future<Output = ()> + Send>>`

Reply:
That did the trick! Thanks a lot!